opentelemetry-contrib/src/datadog/model.rs

```diff
@@ -0,0 +1,202 @@
+use crate::datadog::intern::StringInterner;
+use opentelemetry::api::{Key, Value};
+use opentelemetry::exporter::trace;
+use rmp::encode::ValueWriteError;
+use std::fmt;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum Error {
+    MessagePackError,
+}
+
+impl std::error::Error for Error {}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::MessagePackError => write!(f, "message pack error"),
+        }
+    }
+}
+
+impl From<rmp::encode::ValueWriteError> for Error {
+    fn from(_: ValueWriteError) -> Self {
+        Self::MessagePackError
+    }
+}
+
+// Protocol documentation sourced from https://github.com/DataDog/datadog-agent/blob/c076ea9a1ffbde4c76d35343dbc32aecbbf99cb9/pkg/trace/api/version.go
+//
+// The payload is an array containing exactly 2 elements:
+//
+// 	1. An array of all unique strings present in the payload (a dictionary referred to by index).
+// 	2. An array of traces, where each trace is an array of spans. A span is encoded as an array having
+// 	   exactly 12 elements, representing all span properties, in this exact order:
+//
+// 		 0: Service   (uint32)
+// 		 1: Name      (uint32)
+// 		 2: Resource  (uint32)
+// 		 3: TraceID   (uint64)
+// 		 4: SpanID    (uint64)
+// 		 5: ParentID  (uint64)
+// 		 6: Start     (int64)
+// 		 7: Duration  (int64)
+// 		 8: Error     (int32)
+// 		 9: Meta      (map[uint32]uint32)
+// 		10: Metrics   (map[uint32]float64)
+// 		11: Type      (uint32)
+//
+// 	Considerations:
+//
+// 	- The "uint32" typed values in "Service", "Name", "Resource", "Type", "Meta" and "Metrics" represent
+// 	  the index at which the corresponding string is found in the dictionary. If any of the values are the
+// 	  empty string, then the empty string must be added into the dictionary.
+//
+// 	- None of the elements can be nil. If any of them are unset, they should be given their "zero-value". Here
+// 	  is an example of a span with all unset values:
+//
+// 		 0: 0                    // Service is "" (index 0 in dictionary)
+// 		 1: 0                    // Name is ""
+// 		 2: 0                    // Resource is ""
+// 		 3: 0                    // TraceID
+// 		 4: 0                    // SpanID
+// 		 5: 0                    // ParentID
+// 		 6: 0                    // Start
+// 		 7: 0                    // Duration
+// 		 8: 0                    // Error
+// 		 9: map[uint32]uint32{}  // Meta (empty map)
+// 		10: map[uint32]float64{} // Metrics (empty map)
+// 		11: 0                    // Type is ""
+//
+// 		The dictionary in this case would be []string{""}, having only the empty string at index 0.
+//
+pub(crate) fn encode_v05(
```

Comment:
Might be better to make this an enum if we are expecting multiple version. But also OK if do it in a separate PR.

Reply:
Added - and implemented the v0.3 API whilst I was at it